Archive/Wiedijk100Theorems/BuffonsNeedle.lean

```diff
@@ -0,0 +1,394 @@
+import Mathlib.Analysis.SpecialFunctions.Integrals
+import Mathlib.Probability.Distributions.Uniform
+
+/-!
+
+# Freek № 99: Buffon's Needle
+
+This file proves Theorem 99 from the [100 Theorems List](https://www.cs.ru.nl/~freek/100/), also
+known as Buffon's Needle, which gives the probability of a needle of length `l > 0` crossing any
+one of infinite vertical lines spaced out `d > 0` apart.
+
+The two cases are proven in `buffon_short` and `buffon_long`.
+
+## Overview of the Proof
+
+We define a random variable `B : Ω → ℝ × ℝ` with a uniform distribution on `[-d/2, d/2] × [0, π]`.
+This represents the needle's x-position and angle with respect to a vertical line. By symmetry, we
+need to consider only a single vertical line positioned at `x = 0`. A needle therefore crosses the
+vertical line if its projection onto the x-axis contains `0`.
+
+We define a random variable `N : Ω → ℝ` that is `1` if the needle crosses a vertical line, and `0`
+otherwise. This is defined as `fun ω => Set.indicator (needleProjX l (B ω).1 (B ω).2) 1 0`.
+f
+As in many references, the problem is split into two cases, `l ≤ d` (`buffon_short`), and `d ≤ l`
+(`buffon_long`). For both cases, we show that
+```
+ℙ[N] = (d * π) ⁻¹ *
+    ∫ θ in 0..π,
+      ∫ x in Set.Icc (-d / 2) (d / 2) ∩ Set.Icc (-θ.sin * l / 2) (θ.sin * l / 2), 1
+```
+In the short case `l ≤ d`, we show that `[-l * θ.sin/2, l * θ.sin/2] ⊆ [-d/2, d/2]`
+(`short_needle_inter_eq`), and therefore the inner integral simplifies to
+```
+∫ x in (-θ.sin * l / 2)..(θ.sin * l / 2), 1 = θ.sin * l
+```
+Which then concludes in the short case being `ℙ[N] = (2 * l) / (d * π)`.
+
+In the long case, `l ≤ d` (`buffon_long`), we show the outer integral simplifies to
+```
+∫ θ in 0..π, min d (θ.sin * l)
+```
+which can be expanded to
+```
+2 * (
+  ∫ θ in 0..(d / l).arcsin, min d (θ.sin * l) +
+  ∫ θ in (d / l).arcsin..(π / 2), min d (θ.sin * l)
+)
+```
+We then show the two integrals equal their respective values `l - (l^2 - d^2).sqrt` and
+`(π / 2 - (d / l).arcsin) * d`. Then with some algebra we conclude
+```
+ℙ[N] = (2 * l) / (d * π) - 2 / (d * π) * ((l^2 - d^2).sqrt + d * (d / l).arcsin) + 1
+```
+
+## References
+
+* https://en.wikipedia.org/wiki/Buffon%27s_needle_problem
+* https://www.math.leidenuniv.nl/~hfinkeln/seminarium/stelling_van_Buffon.pdf
+* https://www.isa-afp.org/entries/Buffons_Needle.html
+
+-/
+
+open MeasureTheory (MeasureSpace IsProbabilityMeasure Measure pdf.IsUniform)
+open ProbabilityTheory Real
+
+lemma set_integral_toReal_ofReal_nonneg_ae {α : Type*} [MeasureSpace α] {s : Set α} {f : α → ℝ}
+    (hs : MeasurableSet s) (hf : ∀ᵐ x : α, x ∈ s → 0 ≤ f x) :
+    ∫ (x : α) in s, ENNReal.toReal (ENNReal.ofReal (f x)) =
+    ∫ (x : α) in s, f x := by
+
+  have : ∀ᵐ x : α, x ∈ s → f x = ENNReal.toReal (ENNReal.ofReal (f x)) := by
+    simp_rw [eq_comm (a := f _), ENNReal.toReal_ofReal_eq_iff, hf]
+
+  rw [MeasureTheory.set_integral_congr_ae hs this]
```

Comment:
Can you inline this lemma?

Reply:
I definitely don't like how it's at the top of the file, but it's used both in `buffon_short` and `buffon_long`.

Reply:
Ok I've inlined it